src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,331 @@
+            ECPoint[] notaries = null;
+            foreach (Transaction tx in engine.PersistingBlock.Transactions)
+            {
+                var attr = tx.GetAttribute<NotaryAssisted>();
+                if (attr is not null)
+                {
+                    notaries ??= GetNotaryNodes(engine.SnapshotCache);
+                    var nKeys = attr.NKeys;
+                    nFees += (long)nKeys + 1;
+                    if (tx.Sender == Hash)
+                    {
+                        var payer = tx.Signers[1];
+                        var balance = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, payer.Account))?.GetInteroperable<Deposit>();
+                        balance.Amount -= tx.SystemFee + tx.NetworkFee;
+                        if (balance.Amount.Sign == 0) RemoveDepositFor(engine.SnapshotCache, payer.Account);
+                    }
+                }
+            }
+            if (nFees == 0) return;
+            var singleReward = CalculateNotaryReward(engine.SnapshotCache, nFees, notaries.Length);
+
+            foreach (var notary in notaries) await GAS.Mint(engine, notary.EncodePoint(true).ToScriptHash(), singleReward, false);
+        }
+
+        protected override void OnManifestCompose(IsHardforkEnabledDelegate hfChecker, uint blockHeight, ContractManifest manifest)
+        {
+            manifest.SupportedStandards = new[] { "NEP-27" };
+        }
+
+        /// <summary>
+        /// Verify checks whether the transaction is signed by one of the notaries and
+        /// ensures whether deposited amount of GAS is enough to pay the actual sender's fee.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="sig">Signature</param>
+        /// <returns>Whether transaction is valid.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private bool Verify(ApplicationEngine engine, byte[] sig)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            if (tx.GetAttribute<NotaryAssisted>() is null) return false;
+            foreach (var signer in tx.Signers)
+            {
+                if (signer.Account == Hash)
+                {
+                    if (signer.Scopes != WitnessScope.None) return false;
+                    break;
+                }
+            }
+            if (tx.Sender == Hash)
+            {
+                if (tx.Signers.Length != 2) return false;
+                var payer = tx.Signers[1].Account;
+                var balance = GetDepositFor(engine.SnapshotCache, payer);
+                if (balance is null || balance.Amount.CompareTo(tx.NetworkFee + tx.SystemFee) < 0) return false;
+            }
+            ECPoint[] notaries = GetNotaryNodes(engine.SnapshotCache);
+            var hash = tx.GetSignData(engine.GetNetwork());
+            return notaries.Any(n => Crypto.VerifySignature(hash, sig, n));
```

Comment:
How much notary nodes will exists? the cost of this method should depend of it

Reply:
As many as council wants. 4 is not a bad number, the cost is rather predictable.

Reply:
NotaryNode role follows native Designation rules, so right now the standard constraint is 32.